### DISCUSSION TITLE
Client Policies vs Client Configuration

### DISCUSSION THREAD

**Discussion by tnorimat** [keycloak/keycloak#8558] [Ideas]
### Preface
This topic has already been discussed in [keycloak-dev mailing list's thread](https://groups.google.com/g/keycloak-dev/c/U8m3VirOHjk/m/vqhR1hbxBwAJ).

By following the updated [keycloak's contribution guide](https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md), I've post this topic here.

Also, as [the keycloak's contribution guide](https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md#open-a-discussion-on-a-proposed-change) described, more low-level implementation discussions continue in the keycloak-dev mailing list's thread.

### Discussion point

Basically, keycloak executes logics on a request from a client by following its client settings(metadata).
E.g., authorization request on authorization Endpoint, token request on token endpoint.

Current client policies' built-in executors cannot override such logics. 

### Proposal

Revising client policies to override logics based on client settings when keycloak receives a request from a client.
Considering the nature of client "policies", it is better to precede client setting based logics.

### Motivation

- Security
It can enforce logics satisfying some security level regardless of client settings.

- Flexibility
It can change such enforcing logics dynamically per request from a client

### Use Case

- Apply different FAPI security profile per request from the same client dynamically

FAPI 1.0 Baseline security profiles are applied on some requests from a client while FAPI 1.0 Advanced security profile are applied on others from the same client.

The existing client policies can support this use case. However, there is the case that some request obeying FAPI 1.0 Baseline is not accepted.

### Notes

This discussion point has been treated in [FAPI-SIG](https://github.com/keycloak/kc-sig-fapi). Its detail can be found in [the FAPI-SIG's discussion document](https://github.com/keycloak/kc-sig-fapi/blob/master/FAPI-SIG/documents/ClientPolicies/ClientPoliciesRevised.pdf).

**Comment by stianst** [comment]
Would we also want to consider deprecating/removing some of the options at the client level?

**Reply by tnorimat** [reply]
@stianst We can remove/deprecate some existing client settings. IMO, client metadata defined some standard documents (E.g. OAuth2/OIDC dynamic client registration) should not be deprecated/removed basically to follow these standards.

**Reply by stianst** [reply]
Sounds good
